view/src/app/decorators/auto-unsub.ts

```diff
@@ -10,7 +10,7 @@ import { Subscription } from "rxjs";
  */
 export function AutoUnsub() {
     return function (constructor: any) {
-        const orig = constructor.prototype.ngOnDestroy
+        const orig = (constructor.prototype.ngOnDestroy && constructor.prototype.ngOnDestroy.name === 'ngOnDestroy') ?  constructor.prototype.ngOnDestroy : undefined;
```

Comment:
i componenti che hanno delle subscriptions ereditano dalla classe che funge da gestore delle subscriptions, questa classe è annotata con AutoUnsub. A startup quindi inizialmente `orig` è `undefined` perché `constructor.prototype` fa riferimento alla classe `super`. Viene quindi creato un metodo `ngOnDestroy` sulla classe padre che serve a gestire le unsubscribe. Le classi che ereditano questa classe padre ereditano anche l'annotazione. Nelle successive resolve dell'annotazione `constructor.prototype` fa riferimento ad uno dei componenti specifici, questi ereditano la definizione del padre della funzione  `ngOnDestroy` , questa però altro non è che una funzione anonima creata con `constructor.prototype.ngOnDestroy = function ()`, di conseguenza chiamando `orig.apply()` richiameremmo la funzione anonima dove lo scope `this` fa riferimento alla funzione e non al componente. La soluzione proposta prevede di distinguere i casi

- `constructor.prototype.ngOnDestroy.name === 'ngOnDestroy'` in questo caso il componente specifico ha effettivamente una sua definizione del metodo `ngOnDestroy`
- `else` il metodo `ngOnDestroy` del componente specifico non esiste e quindi non deve essere chiamato

Una soluzione più sporca e non testata sarebbe quella di modificare la funzione da `function ()` a `()=>` in questo modo lo scope dovrebbe essere mantenuto, ma chiameremmo due volte le unsubscribe